Allow services with no functions #2267

johncmckim opened the pull request:
## What did you implement:
Closes #2267

Allows people to create services that do not contain functions.

## How did you implement it:
Simply removed the error and defaulted the functions property to an empty object.

## How can we verify it:
Create a project with a simple config and try a deployment.

```
service: test-sls-proj
provider:
  name: aws
  runtime: nodejs4.3
```

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below


***Is this ready for review?:*** YES

johncmckim commented:
I haven't updated documentation because I couldn't see anything explicitly saying `functions` was required. But if you want me to add documentation somewhere I'm happy to do so.

nicka commented:
Thank you for this 👍

eahefnawy commented:
@johncmckim We just merged [change to policy statement resources](https://github.com/serverless/serverless/pull/2520). As a result, this PR fails when you have the following config:

```yml
provider:
  cfLogs: true
```
This is because when there are no functions in your service and you try to give rights to specific log groups that doesn't exist, the Resources array in the policy document is empty, which is invalid in CF world.

Generally speaking though, if there are no functions, there shouldn't be any Lambda IAM roles/policies imo. What you think @johncmckim ? cc/ @flomotlik

nicka commented:
>Generally speaking though, if there are no functions, there shouldn't be any Lambda IAM roles/policies imo.

Agreed!

johncmckim commented:
@eahefnawy I rebased, ran the tests and tried a deploy. But I couldn't produce an error related to `cfLogs: true`

However, I do agree that we shouldn't create unnecessary IAM roles. So I have just updated the PR to skip adding the IAM roles if there are no functions.

eahefnawy commented:
Thanks @johncmckim 😊 ... this makes deployment much faster because creating IAM resources takes most of the deployment time.

G2M! 😊

nicka commented:
Thanks @johncmckim @eahefnawy, super useful(is now updating a few stacks).
